bisection.go

```diff
@@ -60,18 +56,7 @@ func (b *Bisection) Init(initLoc LinesearchLocation, initStepSize float64, f *Fu
 	return FunctionAndGradientEval
 }
 
-const (
-	funcSmallEqual = 1e-14
-	gradSmallEqual = 1e-10
-)
-
 func (b *Bisection) Finished(l LinesearchLocation) bool {
-	if floats.EqualWithinRel(l.F, b.initF, funcSmallEqual) && math.Abs(l.Derivative) < gradSmallEqual && math.Abs(b.initGrad) < gradSmallEqual {
-		// The two numbers are so close that we should just check on the gradient
-		// TODO: Should iterate be updated? Maybe find a function where it needs it.
-		return math.Abs(l.Derivative) < b.GradConst*math.Abs(b.initGrad)
-	}
-
 	return StrongWolfeConditionsMet(l.F, l.Derivative, b.initF, b.initGrad, b.currStep, 0, b.GradConst)
```

Comment:
Strictly, the strong Wolfe conditions only hold if FunConst > 0. However, it can be any number greater than zero, including very small values. In finite precision world, when we get close to the minimum, there are large regions where the function value is "constant" as far as representable float64 are concerned. In this region, the line search has to fail because there are no locations with a smaller function value, even though with infinite precision there would be many steps that satisfy the condition. Setting C1 to zero avoids those problems. On a conceptual level, we care about the function not increasing and the gradient getting smaller. Since these methods assume smooth and continuous functions, driving the gradient to zero is the same as finding the location with minimum F. We're happy stopping the linesearch with a sufficient decrease in gradient norm.